task-6: Implement CRUD API for resource service

ityaptin opened the pull request:
- CRUD service with following methods: CreateEntity, DescribeEntity, ListEntities, RemoveEntity
- proto files for grpc and http server generation
- makefile and dockerfile for build
- start grpc and http servers in cmd/ocp-resource-api

ityaptin committed:
task-6: Implement CRUD API for resource service

ityaptin committed:
task-7: Add repo calls to the API implementation

ityaptin committed:
task-8: Implement task 8

This commit implements following:
- Add Multi Create endpoint
- Add Update endpoint
- Trace multi create endpoint]
- Add CUD request metrics for prometheus
- Send action notifications to kafka

2nick commented on api/ocp-resource-api/ocp-resource-api.proto:
```diff
@@ -0,0 +1,114 @@
+}
+
+message DescribeResourceRequestV1 {
+  uint64 resource_id = 1;
```
несколько путает, что где-то id, а где-то resource_id - хотелось бы чего-то одного

ityaptin replied:
Fixed

2nick commented on cmd/ocp-resource-api/main.go:
```diff
@@ -1,38 +1,129 @@
 package main
 
 import (
```
Обычно импорт форматируется так, что стандартные пакеты идут одним списком сверху и через пустую линию внешние модули, всё сортируется по алфавиту.

Вообще Goland такое обычно делает при форматировании, ещё можно использовать goimports.

ityaptin replied:
Fixed

2nick commented on internal/api/request.go:
```diff
@@ -0,0 +1,129 @@
+		return nil, status.Error(codes.Internal, fmt.Sprint("flush err"))
+	}
+	rsp := desc.MultiCreateResourceResponseV1{}
+	for _ = range resources {
```
можно не писать _ =

ityaptin replied:
Fixed

2nick commented on internal/flusher/flusher.go:
```diff
@@ -25,7 +28,13 @@ func NewFlusher(
 
-func (f *flusher) Flush(resources []models.Resource) []models.Resource {
+func (f *flusher) Flush(ctx context.Context, resources []models.Resource, parentSpan opentracing.Span) []models.Resource {
+	var span opentracing.Span
```
parentSpan обычно достаётся из контекста, а не передаётся как отдельный аргумент

ityaptin replied:
Fixed

2nick commented on internal/producer/producer.go:
```diff
@@ -0,0 +1,53 @@
+		Key:       sarama.StringEncoder(p.topic),
+		Value:     sarama.StringEncoder(bytes),
+		Partition: -1,
+		Timestamp: time.Time{},
```
Зачем инициализировать - поле и так будет заполнено дефолтным значением

ityaptin replied:
Fixed

2nick commented on internal/repo/repo.go:
```diff
@@ -1,9 +1,16 @@
-	DescribeEntity(entityId uint64) (*models.Resource, error)
+	AddEntity(ctx context.Context, entity models.Resource) (*models.Resource, error)
+	AddEntities(ctx context.Context, entities []models.Resource, span opentracing.Span) error
+	ListEntities(ctx context.Context, limit uint64, offset uint64) (*[]models.Resource, error)
```
Обычно возвращают просто слайс, потому что с ссылкой на слайс приходится делать дополнительную проверку на nil

ityaptin replied:
Fixed

2nick commented on internal/repo/repo_pg.go:
```diff
@@ -0,0 +1,209 @@
+	if err != nil {
+		return nil, err
+	}
+	id := uint64(0)
```
обычно переменные, которые инициализируются дефолтным значением задают через var, а не присваиванием

ityaptin replied:
Fixed

2nick commented on internal/tracer/tracer.go:
```diff
@@ -0,0 +1,40 @@
+	// and github.com/uber/jaeger-lib/metrics respectively to bind to real logging and metrics
+	// frameworks.
+	jLogger := jaegerlog.StdLogger
+	jMetricsFactory := metrics.NullFactory
```
можно было бы прикрутить zerolog и метрики :)

ityaptin committed:
task-8: Fix imports format and add some adjusments

- fix default variable init
- return slice instead of slice ptr
- remove redundant timestamp init
- add span to the ctx instead of arg
